github/rules.go

```diff
@@ -651,8 +654,10 @@ func (r *RepositoryRulesetRules) MarshalJSON() ([]byte, error) {
 }
 
 // marshalRepositoryRulesetRule is a helper function to marshal a ruleset rule.
-func marshalRepositoryRulesetRule(t RulesetRuleType, params any) ([]byte, error) {
-	if params == nil {
+func marshalRepositoryRulesetRule[T any](t RulesetRuleType, params T) ([]byte, error) {
+	paramsType := reflect.TypeFor[T]()
```

Comment:
We try to avoid using the `reflect` package in non-test functions in this repo, but I see here where it makes it much nicer for the end user, so let's roll with it and see if we get any feedback.

One way around it, though, might be to write another generator that runs when `./script/generate.sh` is called which will iterate over all the possibilities and then perform a switch with a type cast and call the appropriate code.

If you agree, then maybe you could add a `// TODO: Use generator here instead of reflection` and open up an issue for you or someone else to implement in the future.  If you disagree, I would love to hear why you would want to stick with reflection.

Reply:
I don't specifically disagree, but given we're talking about JSON marshaling (I'll try and remember to leave out the second l) where reflection is already being used I'm not sure what the issue is? Do you have some specific concerns?

FYI I had a working version using generics but it only worked with the pointer having been dereferenced (`new(T)` as the nil check). But I didn't think it was as easy to understand as the reflection version.

Reply:
Oh sorry, I forgot to explain _why_.  We (in this repo, and I, in general) tend to avoid reflection for production code whenever possible due to the possibly-significant performance hit it may incur.

So it might be more appropriate to add two comments here:

```
// TODO: benchmark the code that uses reflection
// TODO: Use a generator here instead of reflection if there is a significant performance hit
```

Reply:
Done